### PR TITLE
Features/receipt form

### LINKED ISSUE #39: Form for Image Scan
### Issuer:
Hung Ngoos
### Assignee(s):
Vien Nguyen

### Requirements:
- Create a form for adjusting the receipts that we have scanned.
- Have the UI like the following pics:
image

For the quantity which is a number input you could refer to the Cart.tsx components
The backend will send a data like the following: 

```typescript
{
    "message": "Image processed successfully",
        "data": {
        "items": [
            {
                "name": "Lorem ipsum",
                "price": "$9.20",
                "quantity": "1"
            },
            {
                "name": "Lorem ipsum dolor sit",
                "price": "$19.20",
                "quantity": "1"
            },
            {
                "name": "Lorem ipsum dolor sit amet",
                "price": "$15.00",
                "quantity": "1"
            },
            {
                "name": "Lorem ipsum",
                "price": "$15.00",
                "quantity": "1"
            },
            {
                "name": "Lorem ipsum dolor sit",
                "price": "$15.00",
                "quantity": "1"
            },
            {
                "name": "Lorem ipsum",
                "price": "$19.20",
                "quantity": "1"
            }
        ],
            "total": "$107.60"
    }
}
```
Therefore, this form will render dynamically based on the inputs not static like register which has only 3 inputs. 
But first, just use mockdata to create this form statically. 
In this issue, you will need to create zod validation linking it with react hook form
This will also use useFieldArray of react hook form to render this array of element. You could also refer to Cart.tsx
The zod validation will require you to create an interface or so-called schema. Based on the mockdata above, create the schema for it.  

### Dependencies:
- zod
- react hook form 

### Done requirements:
-Be able to render dynamically, have the form validated according to the data schema provided by zod.

### PR DESCRIPTION
**Pull request by viennguyen24**
**Issue #39:** Form for image scan and schema validation for user register / login page

**Branch name:** features/receipt-form

**PR #39:** Implemented UI for receipt-form, display mock receipt data and allows user to manually change, update, and delete data. Changed hand-written form validation into schema validation with Zod for user register / login

**Libraries:** React-hook-form, zod, lucide icon react

**Mock data**: Please refer to #47  for mock data

**Demo receipt form**
[https://drive.google.com/file/d/1EExC536tGbBr4BF4lFR29h0NkUIm2Xw6/view?usp=drive_link](url)

**Files involved: client/src/pages/ReceiptForm.tsx**

**Overview**
This is a form page where users can receive data about scanned ingredients and manually update the ingredients / add new ones. 
1. Backend sends data about ingredients that users scanned from their receipt. Each item in the data includes:
- Name of the ingredient
- Price 
- Quantity 
- Suggested ingredients (in case the ingredient was mismatched or irrelevant)

2. User can manually adjust the name / price / quantity in each input field or delete the ingredient by clicking the **Trash icon** button. If the ingredient was mismatched or irrelevant (for example the ingredient was chicken breast but the name was only chicken), then user can click on the **Pencil icon** button where they will receive a list of ingredients that might be the right one.

3. Once all ingredients are selected, users can click on submit changes and send the ingredients back to the backend. The data should look like this: 
```typescript
{receipts: 
     [{name: 'pumpkin vanilla', quantity: '1', price: '165.00'},
      {name: 'tomato & cheese', quantity: '1', price: '185.00' },]
}
```

**Work to be done**
- [ ] Calling backend to retrieve receipts data 
- [x] Creating a state to manage current receipts and load current receipts, as well as to store data retrieved
- [ ]  Break down some code into separate component to reuse in the future (if needed) 
- [ ] Bring hard-fixed data / types outside for reusability
- [ ] Update UI and implement responsiveness / accessibility

**Schema validation for authentication page**
```typescript

\\\ Register.tsx
const UserRegister = z.object({
  username: z.string().min(4, { message: "Username must be at least 4 characters" }),
  email: z.string().email({ message: "Please enter a valid email"}),
  password: z.string().min(8, { message: "Password must be at least 8 characters" }),
  reEnterPassword: z.string().min(8, { message: "Password must be at least 8 characters" }),
}).refine(data => data.password === data.reEnterPassword, {
  message: "You must re-enter your password correctly",
  path: ["reEnterPassword"], 
});

const RequestRegister = z.object({
  name: z.string(),
  email: z.string(),
  password: z.string(),
});

type UserRegisterType = z.infer<typeof UserRegister>;
type RequestRegisterType = z.infer<typeof RequestRegister>;
```

```typescript
\\Login.tsx

const UserLogin = z
  .object({
    email: z.string().email({ message: 'Please enter a valid email' }),
    password: z.string().min(8, { message: 'Please enter a password with 8 characters minimum' }),
  })
  .strict().required({ email: true, password: true });
type UserLoginType = z.infer<typeof UserLogin>;
```

### CODE REVIEW

**Review by hungngodev — CHANGES_REQUESTED** [review]
Change the login to Login file

File: client/src/pages/mockReceipt.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      price: '$15.00',
+      quantity: '1',
+    },
+    {
+      name: 'Lorem ipsum',
+      price: '$19.20',
+      quantity: '1',
+    },
+  ],
+};
```
Comment (hungngodev) [inline client/src/pages/mockReceipt.ts]: Move this mock file into the ReceiptForm.tsx. Also the Receipt Form should be placed in the component folder not pages folder.

File: client/src/pages/login.tsx [hunk 2]
```diff
@@ -9,25 +9,29 @@ import {
...
-interface APIData {
-  email: string;
-  password: string;
-}
+const User = z
+  .object({
+    email: z.string().email({ message: 'Please enter a valid email' }),
+    password: z.string().min(8, { message: 'Please enter a password with 8 characters minimum' }),
+  })
+  .strict();
```
Comment (hungngodev) [inline client/src/pages/login.tsx]: I have not seen the same for the Register page. Could you upload to this same PR?

File: client/src/pages/MockReceipt.ts
Comment (hungngodev) [inline client/src/pages/MockReceipt.ts:1]: Could you move this file with the ReceiptForm.tsx to the components folder?

**Review by hungngodev — CHANGES_REQUESTED** [review]

File: client/src/pages/ReceiptForm.tsx
Comment (viennguyen24) [inline client/src/pages/ReceiptForm.tsx:1]: This file is moved into components/ingredients folder

File: client/src/pages/login.tsx (hunk 2 above)
Comment (viennguyen24) [reply client/src/pages/login.tsx]: Hi Hung, I uploaded both the Register and Login page at commit #[b2e3a42](https://github.com/hungngodev/Flavorie/pull/49/commits/b2e3a427abeaa7b2ec569b5c77b5d93dd927ff45)!

File: client/src/pages/MockReceipt.ts
Comment (viennguyen24) [reply client/src/pages/MockReceipt.ts:1]: > Could you move this file with the ReceiptForm.tsx to the components folder?

Check commit #[7820fba](https://github.com/hungngodev/Flavorie/pull/49/commits/7820fba3cfa249047c060da06faae58a18eb0da9) !

**Comment by hoatuyet423** [comment]
- Confused, is this branch suppose to resolve 2 issues? Only one is mentioned in the description. If yes, then should not do this in the future, only 1 PR per 1 branch per issue. 
- A lot of styling in the code is really inconsistent. For example how your css is structures is different in different files and places in your code. 
- I didnt read much about the html as it will mostly shown in the UI, but leave a few comments here and there
- Can you provide a video of interaction with the UI and what it does?
- NIT: I feel like the huge space in between the item card is making it feel empty. Can make the order summary card bigger to balance this out 
- Overall UI look nice and clean, but feel like the code is still pretty confused to follow .-. Good works tho :>

**Comment by hoatuyet423** [comment]
- resolved conflict 
- demo link doesnt work

**Pull request closed without merging**